torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +856,335 @@ def _get_ignored_params(
+            )
+
+    def _move_module_to_device(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+        device_id: Optional[Union[int, torch.device]],
+    ):
+        """
+        Moves ``module`` depending on ``device_id`` and its current device,
+        where ``device_id`` should be what the user passed into the FSDP
+        constructor. Since the entire module is moved, ignored parameters are
+        moved as well.
+
+        - If ``device_id`` is not ``None``, then this moves ``module`` to the
+        corresponding device.
+        - If ``device_id`` is ``None``, then this does not move ``module`` but
+        warns the user if it is only CPU.
+
+        Precondition: ``_check_single_device_module()``.
+        """
+        cpu_device = torch.device("cpu")
+        param = next(self._get_orig_params(module, ignored_params), None)
+        if param is None:
+            return  # no original parameters to manage
+        if device_id is not None:
+            assert param.device == cpu_device or param.device == self.compute_device, (
+                f"Unexpected device {param.device} -- check the FSDP implementation"
+            )
+            if param.device == cpu_device:
+                # NOTE: This includes moving ignored parameters.
+                module = module.to(device_id)
+                # TODO (awgu): This is a temporary hack to move already-
+                # constructed `FlatParameter`s back to CPU if needed.
+                for submodule in module.modules():
+                    if (
+                        isinstance(submodule, FullyShardedDataParallel)
+                        and submodule.cpu_offload.offload_params
+                    ):
+                        with torch.no_grad():
+                            for handle in submodule._handles:
+                                handle._flat_param_to(torch.device("cpu"))
+        elif param.device == cpu_device:
+            self._warn_cpu_init()
+
+    def _warn_cpu_init(self):
+        warnings.warn(
+            "Module is put on CPU and will thus have flattening and sharding"
+            " run on CPU, which is less efficient than on GPU. We recommend passing in "
+            "`device_id` argument which will enable FSDP to put module on GPU device,"
+            " module must also be on GPU device to work with `sync_module_states=True` flag"
+            " which requires GPU communication."
+        )
+
+    def _sync_module_states(
+        self, module: nn.Module, params: List[nn.Parameter]
+    ) -> None:
+        """
+        Synchronizes module states (i.e. parameters ``params`` and all
+        not-yet-synced buffers) by broadcasting from rank 0 to all ranks.
+        Precondition: ``sync_module_states == True`` and ``self.process_group``
+        has been set.
+        """
+        if params and any(param.device == torch.device("cpu") for param in params):
+            raise ValueError(
```

Comment:
would I be correct that there is no way that users can reach there through FSDP public APIs as the CPU module should have been moved to GPU at this point?

Reply:
this can still happen if user forces a CPU init by passing in a CPU module and no device_id but `sync_module_states=True` which requires GPU comm